tests/ui/transmute.rs

```diff
@@ -152,11 +152,17 @@ fn transmute_ptr_to_ptr() {
         let _: &f32 = std::mem::transmute(&1u32);
         let _: &mut f32 = std::mem::transmute(&mut 1u32);
     }
-    // These should be fine
+    // These should be fine:
+    // Recommendations for solving the above; if these break we need to update
+    // those suggestions
     let _ = ptr as *const f32;
     let _ = mut_ptr as *mut f32;
     let _ = unsafe { &*(&1u32 as *const u32 as *const f32) };
     let _ = unsafe { &mut *(&mut 1u32 as *mut u32 as *mut f32) };
+    // This is just modifying the lifetime, and is one of the recommended uses
+    // of transmute
+    let n = 1u32;
+    let _ = unsafe { std::mem::transmute::<&'_ u32, &'static u32>(&n) };
```

Comment:
Do we want `transmute::<Foo<'a>, Foo<'b>>` to also not lint?

Reply:
I think that's a good question. Are there reasonable cases where one of the transmuted lifetimes is not `'static`? My suspicion is that the answer is yes but I've never run into one.

@sgrif, since you filed the original issue, do you have an opinion here?

Reply:
Yes, there's definitely use cases where neither lifetime is `'static`. The main use case that I hit is literally `Foo<'a>` to `Foo<'b>` where `'b` is known to be smaller than `'a`, but is also unnameable. If `Foo` is invariant, this is an important use case.

Reply:
In all relevant cases for Diesel, `'static` is not involved. :smile:

https://github.com/diesel-rs/diesel/blob/a6b00b3cee355a314e3d2c6a83a772fea2fe5738/diesel/src/query_builder/debug_query.rs#L87-L94
https://github.com/diesel-rs/diesel/blob/a6b00b3cee355a314e3d2c6a83a772fea2fe5738/diesel/src/query_builder/ast_pass.rs#L106-L110

Interestingly, this is also our *only* use of `mem::transmute` which I didn't expect.

Reply:
@oli-obk In that case, I think `transmute::<Foo<'a>, Foo<'b>>` should not lint. Do you want me to add a test explicitly checking this?

Reply:
Yes please